core/dbt/include/global_project/macros/materializations/archive/archive.sql

```diff
@@ -94,71 +124,142 @@
+    {%- endset %}
 
-    merged as (
+    {% set transforms -%}
+    current_data as (
 
-      select *, 'update' as {{ adapter.quote('change_type') }} from updates
-      union all
-      select *, 'insert' as {{ adapter.quote('change_type') }} from insertions
+        select
+            {% for col in source_columns %}
+                {{ col.name }} {% if not loop.last %},{% endif %}
+            {% endfor %},
+            {{ archive_get_time() }} as dbt_updated_at,
+            {{ unique_key }} as dbt_pk,
+            {{ archive_get_time() }} as dbt_valid_from,
+            {{ timestamp_column.literal('null') }} as tmp_valid_to
+        from source
+    ),
+
+    archived_data as (
+
+        select
+            {% for col in source_columns %}
+                {{ col.name }},
+            {% endfor %}
+            dbt_updated_at,
+            {{ unique_key }} as dbt_pk,
+            dbt_valid_from,
+            dbt_valid_to as tmp_valid_to
+        from {{ target_relation }}
+
+    ),
 
+    insertions as (
+
+        select
+            current_data.*,
+            {{ timestamp_column.literal('null') }} as dbt_valid_to
+        from current_data
+        left outer join archived_data
+          on archived_data.dbt_pk = current_data.dbt_pk
+        where
+          archived_data.dbt_pk is null
+          or ( {{ update_candidate }} )
+    ),
+
+    updates as (
+
+        select
+            archived_data.*,
+            {{ archive_get_time() }} as dbt_valid_to
+        from current_data
+        left outer join archived_data
+          on archived_data.dbt_pk = current_data.dbt_pk
+        where {{ update_candidate }}
+    ),
+    {%- endset %}
+
+    {%- set hash_components = ['dbt_pk'] %}
+    {%- do hash_components.extend(check_cols) -%}
+    {%- set scd_hash = archive_hash_arguments(hash_components) -%}
+    {{ archive_select_generic(source_sql, target_relation, transforms, scd_hash) }}
+{%- endmacro %}
+
+{# this is gross #}
+{% macro create_empty_table_as(sql) %}
+  {% set tmp_relation = api.Relation.create(identifier=model['name']+'_dbt_archival_view_tmp', type='view') %}
+  {% set limited_sql -%}
+    with cte as (
+      {{ sql }}
     )
+    select * from cte limit 0
+  {%- endset %}
+  {%- set tmp_relation = create_temporary_table(limited_sql, tmp_relation) -%}
 
-    select *,
-        {{ archive_scd_hash() }} as {{ adapter.quote('scd_id') }}
-    from merged
+  {{ return(tmp_relation) }}
 
 {% endmacro %}
 
+
 {% materialization archive, default %}
   {%- set config = model['config'] -%}
 
   {%- set target_database = config.get('target_database') -%}
   {%- set target_schema = config.get('target_schema') -%}
-  {%- set target_table = config.get('target_table') -%}
-
-  {%- set source_database = config.get('source_database') -%}
-  {%- set source_schema = config.get('source_schema') -%}
-  {%- set source_table = config.get('source_table') -%}
+  {%- set target_table = model.get('alias', model.get('name')) -%}
```

Comment:
I have to think about this some more, but I think we really do want archives to look a more like models when possible, because that makes it easier to both implement and reason about their behavior. I do agree on using `target_schema`/`target_database`, but I think the table name matching the block name makes a lot of sense... models will do that in the future when we have model blocks.